File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

const int INF = 999999;
int answer = INF;
vector<bool> visited;

bool check(string lhs, string rhs) {
    int cnt = 0;
    for(int i = 0; i < lhs.length(); i++) {
        if(lhs[i] != rhs[i]) cnt++;
    }
    
    if(cnt == 1) return true;
    else return false;
}

void dfs(string cur, string target, vector<string> &words, int step) {
    if(answer <= step) return;
    if(cur == target) {
        answer = min(answer, step);
        return;
    }
    
    for(int i = 0; i < words.size(); i++) {
        if(check(cur, words[i]) && !visited[i]) {
            visited[i] = true;
            dfs(words[i], target, words, step + 1);
            visited[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {    
    visited = vector<bool> (words.size(), false);
    dfs(begin, target, words, 0);    
    if(answer == INF) answer = 0;
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def convertible(lhs, rhs):
    cnt = 0
    for l, r in zip(lhs, rhs):
        if l != r:
            cnt += 1
    
    return True if cnt <= 1 else False

def solution(begin, target, words):
    if target not in words:
        return 0
    
    answer = 0
    
    is_used = set([begin])
    dq = deque([(begin, 0)])
    
    while dq:
        cur_word, cur_cnt = dq.pop()
        
        if cur_word == target:
            answer = cur_cnt
            break
        
        for word in words:
            if word in is_used:
                continue
            if not convertible(cur_word, word):
                continue
            
            dq.append((word, cur_cnt + 1))
            is_used.add(word)
    
    return answer